libsawtooth/src/journal/publisher/mod.rs

```diff
@@ -15,121 +16,810 @@
-}
+        // If any injected batches were descheduled, reschedule them to ensure they get executed
+        for id in &candidate_block.injected_batch_ids {
+            if let Some(batch) = descheduled_batches.remove(id) {
+                candidate_block.scheduler.add_batch(batch)?;
+            }
+        }
 
-pub trait SyncPublisher: Send + Sync {
-    fn box_clone(&self) -> Box<dyn SyncPublisher>;
+        candidate_block.scheduler.finalize()?;
 
-    fn state(&self) -> &Arc<RwLock<Box<dyn PublisherState>>>;
+        // Collect batch execution results in a `HashMap` indexed by batch ID
+        let mut batch_results = HashMap::new();
+        while let Ok(Some(batch_result)) = candidate_block.batch_results.recv() {
+            batch_results.insert(
+                batch_result.batch.batch().header_signature().to_string(),
+                batch_result,
+            );
+        }
 
-    fn on_chain_updated(
-        &self,
-        state: &mut Box<dyn PublisherState>,
-        chain_head: BlockPair,
-        committed_batches: Vec<Batch>,
-        uncommitted_batches: Vec<Batch>,
-    );
+        // Parse the batch results into batches and state changes
+        let (executed_batches, state_changes) = candidate_block
+            // Get all batch execution results in the order they were originally scheduled in
+            .scheduled_batch_ids
+            .iter()
+            .filter_map(|id| batch_results.remove(id))
+            // Split batches and state changes
+            .fold(
+                (vec![], vec![]),
+                |(mut executed_batches, mut state_changes),
+                 BatchExecutionResult { batch, receipts }| {
+                    for receipt in receipts {
+                        match Vec::try_from(receipt) {
+                            Ok(mut changes) => state_changes.append(&mut changes),
+                            Err(_) => {
+                                debug!(
+                                    "Batch contains invalid transaction, dropping batch: {}",
+                                    batch.batch().header_signature()
+                                );
+                                return (executed_batches, state_changes);
+                            }
+                        }
+                    }
+
+                    if batch.batch().trace() {
+                        trace!("TRACE {}: executed", batch.batch().header_signature());
+                    }
+                    executed_batches.push(batch.take().0);
+
+                    (executed_batches, state_changes)
+                },
+            );
+
+        // Save the batches that were executed and the resulting state root hash of the state
+        // changes
+        candidate_block.executed_batches = executed_batches;
+        candidate_block.state_root_hash = Some(self.merkle_state.compute_state_id(
+            &String::from(candidate_block.previous_block.block().header_signature()),
```

Comment:
This should be the previous block state root hash not signature (also encode in hex).

Reply:
Good catch; fixed